U4-9509 Fix bugs in shadow filesystem

clausjensen opened the pull request.

zpqrtbnk commented on src/Umbraco.Core/IO/ShadowFileSystem.cs:
```diff
@@ -201,21 +201,21 @@ public void AddFile(string path, Stream stream, bool overrideIfExists)
             return files
                 .Except(shadows.Where(kvp => (kvp.Value.IsFile && kvp.Value.IsDelete) || kvp.Value.IsDir)
                     .Select(kvp => kvp.Key))
                 .Union(shadows.Where(kvp => kvp.Value.IsFile && kvp.Value.IsExist).Select(kvp => kvp.Key))
```
I think there's one thing missing here, you are not filtering the files added by the Union statement, which are files that are not in `_fs` (the actual FS) yet but are in the temp FS.

clausjensen commented:
@zpqrtbnk Have fixed up the filtering now and added a bunch of tests to ensure they work as intended. Also added a simple filter test that ensures our filtering function vs. the filtering going on in the .NET framework works the same. Filters in .NET have some broken behavior so we have to match that in the shadow filesystem.
